### PR TITLE
(refactor): Convert dataPlatforms to GMA aspect models and associated resource to GMA resource.

### PR DESCRIPTION
**Pull request by nagarjunakanamarlapudi**
Convert dataPlatforms to GMA aspect models and associated resource to GMA resource.

BREAKING CHANGE: /datasets/dataPlatforms API is now changed to become GMA resource. The schema of the resource also changes.

**What breaks**: Get method of /dataset/dataPlatforms will return com.linkedin.dataPlatforms.DataPlatform schema. Earlier the API returns schema com.linkedin.dataplatform.DataPlatformInfo. 
**How to fix for existing readers**: dataPlatformInfo is a field in then new schema returned by the API. This field should be the same as previous response of the API

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java [hunk 1]
```diff
@@ -1,40 +1,140 @@
...
   @Named("dataPlatformLocalDAO")
-  private ImmutableLocalDAO<DataPlatformAspect, DataPlatformUrn> _localDAO;
+  private BaseLocalDAO<DataPlatformAspect, DataPlatformUrn> _localDAO;
 
+  /**
+   * Get data platform
+   *
+   * @param platformName name of the platform
+   * @param aspectNames list of aspects to be retrieved. Null to retrieve all aspects of the dataPlatforms
+   *
```
Comment (jplaisted) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java:56]: Nit: no empty line between javadoc tags

File: gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java [hunk 2]
```diff
@@ -1,40 +1,140 @@
...
-    return RestliUtils.toTaskFromOptional(() -> _localDAO.get(DataPlatformInfo.class, new DataPlatformUrn(name)));
+  public Task<DataPlatform> get(@Nonnull String platformName,
+      @QueryParam(PARAM_ASPECTS) @Nullable String[] aspectNames) {
+    return super.get(platformName, aspectNames);
   }
 
   @RestMethod.GetAll
-  public Task<List<DataPlatformInfo>> getAll(@Nonnull @PagingContextParam(defaultCount = 100) PagingContext pagingContext) {
-    return Task.value(_localDAO.list(DataPlatformInfo.class, pagingContext.getStart(), pagingContext.getCount()).getValues());
+  public Task<List<DataPlatformInfo>> getAllDataPlatforms(
```
Comment (jplaisted) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java:68]: Please comment.

File: gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java [hunk 3]
```diff
@@ -1,40 +1,140 @@
...
+    return Task.value(
+        _localDAO.list(DataPlatformInfo.class, pagingContext.getStart(), pagingContext.getCount()).getValues());
+  }
+
+  /**
+   * Get the snapshot of data platform
+   *
+   * @param urnString data platform urn
+   * @param aspectNames list of aspects to be returned. null, when all aspects are to be returned.
+   *
```
Comment (jplaisted) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java:79]: Nit: no empty line between javadoc tags

File: gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java [hunk 4]
```diff
@@ -1,40 +1,140 @@
...
-  public Task<List<DataPlatformInfo>> getAll(@Nonnull @PagingContextParam(defaultCount = 100) PagingContext pagingContext) {
-    return Task.value(_localDAO.list(DataPlatformInfo.class, pagingContext.getStart(), pagingContext.getCount()).getValues());
+  public Task<List<DataPlatformInfo>> getAllDataPlatforms(
+      @Nonnull @PagingContextParam(defaultCount = 100) PagingContext pagingContext) {
+    return Task.value(
+        _localDAO.list(DataPlatformInfo.class, pagingContext.getStart(), pagingContext.getCount()).getValues());
+  }
+
+  /**
+   * Get the snapshot of data platform
```
Comment (jplaisted) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java:75]: Nit: complete sentences for Javadoc descriptions (should end with a period).

File: gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java [hunk 5]
```diff
@@ -1,40 +1,140 @@
...
+    super(DataPlatformSnapshot.class, DataPlatformAspect.class);
+  }
 
   @Inject
   @Named("dataPlatformLocalDAO")
-  private ImmutableLocalDAO<DataPlatformAspect, DataPlatformUrn> _localDAO;
+  private BaseLocalDAO<DataPlatformAspect, DataPlatformUrn> _localDAO;
 
+  /**
+   * Get data platform
```
Comment (jplaisted) [inline gms/impl/src/main/java/com/linkedin/metadata/resources/dataplatform/DataPlatforms.java:52]: Nit: complete sentences for Javadoc descriptions (should end with a period).

File: gms/api/src/main/idl/com.linkedin.dataplatform.dataPlatforms.restspec.json [hunk 6]
```diff
@@ -2,7 +2,7 @@
   "name" : "dataPlatforms",
   "namespace" : "com.linkedin.dataplatform",
   "path" : "/dataPlatforms",
-  "schema" : "com.linkedin.dataplatform.DataPlatformInfo",
+  "schema" : "com.linkedin.dataPlatforms.DataPlatform",
```
Comment (jplaisted) [inline gms/api/src/main/idl/com.linkedin.dataplatform.dataPlatforms.restspec.json:5]: This isn't backwards compatible, right? Please

1) consider the consequences for users as to how they will be impacted
2) communicate that clearly in the commit description
3) help achieve 2 by following out commit description standards

https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#footer

**Review by jplaisted — CHANGES_REQUESTED** [review]

File: gms/api/src/main/idl/com.linkedin.dataplatform.dataPlatforms.restspec.json (hunk 6 above)
Comment (nagarjunakanamarlapudi) [reply gms/api/src/main/idl/com.linkedin.dataplatform.dataPlatforms.restspec.json:5]: Updated the commit description to call this out.

Comment (jplaisted) [reply gms/api/src/main/idl/com.linkedin.dataplatform.dataPlatforms.restspec.json:5]: That isn't very descriptive to a lay man. How did it break people specifically? What's the fix?

Comment (nagarjunakanamarlapudi) [reply gms/api/src/main/idl/com.linkedin.dataplatform.dataPlatforms.restspec.json:5]: Take 2 : Updated it again. Let me know if it looks good now.

**Comment by camelliazhang** [comment]
You may want to add migration plan for this API? Do we plan to backfill these existing platforms and remove this json file? https://github.com/linkedin/datahub/blob/125ae288f1601bef2c42f44db6d8eb58b380a304/gms/impl/src/main/resources/DataPlatformInfo.json

**Comment by nagarjunakanamarlapudi** [comment]
> You may want to add migration plan for this API? Do we plan to backfill these existing platforms and remove this json file? https://github.com/linkedin/datahub/blob/125ae288f1601bef2c42f44db6d8eb58b380a304/gms/impl/src/main/resources/DataPlatformInfo.json

Until there is a S&D use case for data platforms, I will like to keep using the json file. Easy to maintain and readable as there won't be too many platforms.

**Review by jplaisted — APPROVED** [review]
Approved, but also get @camelliazhang 's approval too. Thanks

**Comment by camelliazhang** [comment]
> > You may want to add migration plan for this API? Do we plan to backfill these existing platforms and remove this json file? https://github.com/linkedin/datahub/blob/125ae288f1601bef2c42f44db6d8eb58b380a304/gms/impl/src/main/resources/DataPlatformInfo.json
> 
> Until there is a S&D use case for data platforms, I will like to keep using the json file. Easy to maintain and readable as there won't be too many platforms.

Thanks for the reply, Naga. I assume this PR is for this feature? https://github.com/linkedin/datahub/issues/2018
Could you please describe the ingestion plan (write) for dataPlatforms? It sounds like we will still use immutableDAO to write with the content from JSON file?

**Comment by nagarjunakanamarlapudi** [comment]
> > > You may want to add migration plan for this API? Do we plan to backfill these existing platforms and remove this json file? https://github.com/linkedin/datahub/blob/125ae288f1601bef2c42f44db6d8eb58b380a304/gms/impl/src/main/resources/DataPlatformInfo.json
> > 
> > 
> > Until there is a S&D use case for data platforms, I will like to keep using the json file. Easy to maintain and readable as there won't be too many platforms.
> 
> Thanks for the reply, Naga. I assume this PR is for this feature? #2018
> Could you please describe the ingestion plan (write) for dataPlatforms? It sounds like we will still use immutableDAO to write with the content from JSON file?

This PR is actually step 1 for #1956.  

That said, after this we can extend from this state of using ImmutableDAO to using EbeanDAO (#2018)

**Review by camelliazhang — APPROVED** [review]
LGTM

**Pull request merged**